### ISSUE TITLE
Sidekiq Web UI timeouts at ~4-5k connections

### ISSUE THREAD

**Issue by sulleh** [sidekiq/sidekiq#5949]
Ruby version: 2.7.2
Rails version: 6.1.5.1 
Sidekiq / Pro / Enterprise version(s): Sidekiq 6.4.2 / Sidekiq Pro 5.3.1 

We're encountering Sidekiq UI timeouts whenever we have more than 5,000 connections to Redis, and I'm trying to figure out what our best course of action is to address.

I've read many posts from @mperham (and the wiki) and have seen the possibility of 50,000 / sec job throughput, and while we have millions of  jobs in the queue, our workers are headless browsers that are crawling, and most jobs on average take 1-2 mins to complete.  It seems like our scaling problems are based on connections from the Sidekiq processes than anything around. 

1. it's usually around 5,000 connections that our Sidekiq UI hits timeouts. is there an upper bound of Redis connections that Sidekiq can effectively use, or is it entirely limited by CPU / latency?  Our Heroku plan has us supported up to 25,000 connections, so I'm not sure why we are timing out at around 5,000. 

2. if the Sidekiq UI is timing out, how concerned should we be that overall Sidekiq / Worker / Process performance is severely degraded? even though the Sidekiq UI is offline under high load, the Sidekiq processes on the workers are seemingly fine - they  get new jobs from Redis (they only need a new job ever 1-2 mins), but it's hard to know how much of a problem this is (or isn't).

3.  When we have > 5,000 connections, could we just control Sidekiq via the API, or will the API suffer from the same timeouts as the UI? (apologies if thats a stupid question, and the UI isn't in polling mode) 

4. As we have 2-3 Sidekiq processes per instance, I'm not sure if they would benefit from using the [connection_pool](https://github.com/mperham/connection_pool) gem, or if we should focus on migrating off Heroku completely, and just throwing way more CPU at the problem, or just be ok that the UI is offline, or go nuclear and start sharding. 

As our Redis runs on Heroku, we have limited visibility into things like SLOWLOG or anything that requires CONFIG.   I can attach an RTT output if that's helpful - some of the values are high, which I know is bad, but again, the worker performance itself doesn't seem too impacted (apologies if that is a very naive interpretation.) 

We are also open to using Sidekiq Enterprise if that has a feature or functionality that would help, but it pursuing the ent docs nothing jumped out at me.  

**Our Setup** 
- Redis is on Heroku (Premium 9: 25,000 max connections, 10 GB RAM) us-east-1 i think? 
- Core Rails app is also on Heroku, Performance-L Dynos (14GB RAM) 
- Workers run on ARM / x86 EC2 instances, usually on the lower end. 
- EC2 workers running in us-east-1 and us-east-2. 

We utilize two different sidekiq.yml files, depending on the workers that we spin up: 

`sidekiq.yml`
```
:queues:
  - [critical, 20]
  - [private_job, 18]
  - [ml, 10]
  - [stats, 7]
  - [default, 2]
  - [low, 1]
  
:concurrency: 2
:max_retires: 3


development:
  :concurrency: 1
staging:
  :concurrency: 2
production: 
  :concurrency: 2
```

`alt-sidekiq.yml`
```
:queues:
  - [critical, 20]
  - [private_job, 18]
  - [stats, 7]
:concurrency: 4

production:
  :concurrency: 3
```

here's our initializer: 
`sidekiq.rb` 
```
require 'sidekiq'
require 'sidekiq-status'
require "sidekiq/cloudwatchmetrics"

Sidekiq.configure_client do |config|
  config.redis = { :size => 1,  url: ENV["REDIS_URL"], network_timeout: 10, pool_timeout: 5, ssl_params: { verify_mode: OpenSSL::SSL::VERIFY_NONE } }
  Sidekiq::Status.configure_client_middleware config, expiration: 45.minutes
end

Sidekiq.strict_args!(false)

Sidekiq.configure_server do |config|

  config.redis = { :size => 50, url: ENV["REDIS_URL"], network_timeout: 10, pool_timeout: 5, ssl_params: { verify_mode: OpenSSL::SSL::VERIFY_NONE } }
  config.on(:startup) do
    schedule_file = "config/sidekiq_schedule.yml"
    if File.exist?(schedule_file)
      Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)
		end
  end
  # accepts :expiration (optional)
  Sidekiq::Status.configure_server_middleware config, expiration: 45.minutes

  # accepts :expiration (optional)
  Sidekiq::Status.configure_client_middleware config, expiration: 45.minutes

  config.super_fetch!
end

if Rails.env != "development"
  credentials = Aws::Credentials.new(ENV['AWS_EC2_PUBLIC_KEY'], ENV['AWS_EC2_SECRET_KEY'])
  cloudwatch = Aws::CloudWatch::Client.new(region: 'us-east-1', credentials: credentials, logger: nil )
  Sidekiq::CloudWatchMetrics.enable!(client: cloudwatch, namespace: "Sidekiq-#{Rails.env}", process_metrics: false)
end
```

**Comment by mperham** [comment]
The Sidekiq UI is a Rack app. It doesn't know or manage anything about web connections, that's wholly up to your app server, eg puma.

**Comment by sulleh** [comment]
is there something other than connections that would cause the UI to timeout under increased load? Does that mean the /busy route is waiting for a response from all the clients in connection pool via Puma?

**Comment by sulleh** [comment]
Sorry - I may have misspoke.  We have 4000 concurrent connections to the Redis cluster.  We have observed that under that load, the web UI frequently times out.  I suspect there is correlation here between the connection count and CPU usage (we connect via the RedisInsight app to our Heroku Redis instance and see the CPU utilization creep up as the connection count increases.  

So the 4,000 connections are to Redis, and the UI generates timeouts with this load - but our commands to the Sidekiq API via a Heroku console seem pretty speedy, and never get dropped, so i'm trying to understand why does the UI fail / what is the nature of the UI's dependency on Redis, when the API seems to have much less of an issue.

**Comment by mperham** [comment]
Are all UI pages timing out or just a specific one? The only fix that might help was in 6.5.8. At that scale the Busy page becomes quite intensive to render. Can you reproduce it locally by pointing your local web process to the REDIS_URL in Heroku? (security might not allow this)

**Comment by mperham** [comment]
On /busy, the Processes and Jiobs tables are not paginated, they show all data at once. This can be tough with your extreme scale, i.e. many thousands of concurrent jobs executing. It would be reasonable to paginate this with, say, 100 items so the page rendering is predictably O(1), not O(N), but I have no plans to implement this myself. If you want to send a PR, I'd be happy to review.

**Comment by sulleh** [comment]
We aren't using 6.5.8, but we can upgrade and give it a try.  I do think it's likely your last point - the pagination.  It's trying to connect / check in with all the workers and it times out.  We will look into creating a PR for it.  

I'm surprised to hear you describe 4000-5000 connections as "extreme scale" - I know you've described the high end of throughput of 10,000 jobs / sec, but I suppose those could be very small, high speed jobs on a limited number of connections.

**Comment by mperham** [comment]
That's definitely on the larger side. Systems doing 10k+/sec are usually tuned extremely well with many jobs only taking tens of milliseconds. Key is to avoid any Redis operations which are not O(1) and the worker set scan is O(N).

**Comment by sulleh** [comment]
Update - the fix that was in 6.5.8 has definitely improved performance.  The busy page is far more responsive as a result.  We can easily get up to 7500 connections, at which point we are limited by the database. 

We are also seeing another odd behavior, which is that intermittently we see timeouts for the Sidekiq UI (once every couple of hours maybe?) at 4000+ connections, and looking at Redis (KeyDB instance) we are seeing a single core locked at 90%+ during this time. 

I suspect this is some task from KeyDB or Sidekiq pressing the server and locking.  That is out of scope for this thread, so we'll continue investigating on our own.
